### ISSUE TITLE
Populate / `COPY FROM STDIN` and CockroachDB

### ISSUE THREAD

**Issue by anmonteiro** [paurkedal/ocaml-caqti#40]
I've recently ran into some surprising behavior while using [CockroachDB](https://www.cockroachlabs.com/product/) and `populate`.

The symptom I observed was string columns ending up in the database with quotes around them, e.g. `"\"foo\""` instead of `"foo"`.

The cause of this behavior is the escaping implementation in the Postgres driver (CockroachDB uses the PG wire protocol).

While this is certainly a limitation in CockroachDB (it doesn't support the full range of options for COPY, and the same `populate` worked as expected against a Postgres database), I wanted to open this issue to ask if the escaping behavior is necessary when using `populate` calls directly. In a local fork I tried removing them and the `populate` operation worked as expected. 

Wondering what edge cases I missed.

**Comment by paurkedal** [comment]
The current version uses the CSV format. Do you think switching to the TEXT format would solve the issue?

**Comment by anmonteiro** [comment]
Would Caqti need to escape strings in the TEXT format?

**Comment by paurkedal** [comment]
Yes, backslash, newline, carriage return, and tabulator characters will need to be escaped, see "Text Format" in the [the PostgreSQL manual](https://www.postgresql.org/docs/12/sql-copy.html).

**Comment by anmonteiro** [comment]
Got it. I think the `TEXT` format could probably work. Sounds like CockroachDB doesn't support the CSV format.

I'm not sure I can find time to work on this, and it's not very high priority for me as I'm using an `INSERT` statement directly now, but I'm happy to try out a fix whenever anyone finds time to work on it.

From a quick prototype of using `"COPY %s (%s) FROM STDIN WITH TEXT"` and `Param_encoder` (in which `encode_string` is the identity function) it seems to work for my use case.

**Comment by paurkedal** [comment]
In addition to a new `encode_string` we'll need to look at how bytea and NULL are encoded.  NULL should be `\N` and, I may be wrong, but I suspect bytea should be doubly escaped, first with `#escape_bytea` then with the new `encode_string`.

**Comment by paurkedal** [comment]
Can you check if #41 solves the issue?

**Comment by anmonteiro** [comment]
Hey @paurkedal, thanks for getting on this so quickly!

Unfortunately CockroachDB doesn't seem to like the `WITH (FORMAT TEXT)` statement, it just doesn't parse.

Changing it to `WITH TEXT` works for me and solves the issue. Can the option be passed like that to Postgres?

**Comment by paurkedal** [comment]
No, but it's the default, so I pushed a new version without it.

**Comment by anmonteiro** [comment]
Thanks, that is certainly working for my use case now. Note that I haven't done extensive testing, especially around the escaping edge cases.

**Comment by paurkedal** [comment]
I'll merge this for the next release then.  I think it's better to use this format anyway, since it is the native and default format.

**Comment by paurkedal** [comment]
Sorry, I forgot to merge this before releasing. I'll prepare a new release.
